File: 백준/Gold/1379. 강의실 2/강의실 2.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <map>

#define INF 987654321
#define ll long long

using namespace std;

struct Lecture {
    int num;
    ll start, end;
};

struct Room {
    int num;
    ll end;

    bool operator<(const Room &other) const {
        return end > other.end;
    }
};

int N;

bool cmp(const Lecture &a, const Lecture &b) {
    if (a.start == b.start)
        return a.end < b.end;

    return a.start < b.start;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL), cout.tie(NULL);

    cin >> N;

    vector<Lecture> lectures(N);
    for (int n = 0; n < N; ++n)
        cin >> lectures[n].num >> lectures[n].start >> lectures[n].end;

    sort(lectures.begin(), lectures.end(), cmp);

    int k = 1;
    map<int, int> m;
    priority_queue<Room> pq;

    for (auto &lecture: lectures) {
        if (pq.empty()) {
            m[lecture.num] = k;
            pq.push({k++, lecture.end});
        } else {
            if (pq.top().end <= lecture.start) {
                auto curr = pq.top();
                pq.pop();

                m[lecture.num] = curr.num;
                pq.push({curr.num, lecture.end});
            } else {
                m[lecture.num] = k;
                pq.push({k++, lecture.end});
            }
        }
    }

    cout << pq.size() << '\n';
    for (int n = 1; n <= N; ++n)
        cout << m[n] << '\n';

    return 0;
}

File: 백준/Gold/1379. 강의실 2/강의실 2.java
import java.io.*;
import java.lang.*;
import java.util.*;

public class Main {

    private static int N;
    private static int[] Ans;
    private static List<Lecture> Lectures = new ArrayList<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        input(br);

        Collections.sort(Lectures);

        PriorityQueue<Room> pq = new PriorityQueue<>();

        int num = 1;
        for(Lecture lecture: Lectures) {
            if(pq.isEmpty()) {
                Ans[lecture.i] = num;
                pq.add(new Room(num++, lecture.end));
            } else {
                if(pq.peek().time > lecture.start) {
                    Ans[lecture.i] = num;
                    pq.add(new Room(num++, lecture.end));
                } else {
                    Room room = pq.poll();
                    room.time = lecture.end;
                    Ans[lecture.i] = room.num;

                    pq.add(room);
                }
            }
        }

        bw.write(String.valueOf(pq.size()) + '\n');
        for (int i = 1; i <= N; i++) {
            bw.write(String.valueOf(Ans[i]) + '\n');
        }

        br.close();
        bw.close();
    }

    private static void input(BufferedReader br) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        Ans = new int[N + 1];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());

            Lecture lecture = new Lecture();

            lecture.i = Integer.parseInt(st.nextToken());
            lecture.start = Long.parseLong(st.nextToken());
            lecture.end = Long.parseLong(st.nextToken());

            Lectures.add(lecture);
        }
    }

    static class Room implements Comparable<Room> {
        public int num;
        public long time;

        public Room(int num, long time) {
            this.num = num;
            this.time = time;
        }

        @Override
        public int compareTo(Room o) {
            return Long.compare(time, o.time);
        }
    }

    static class Lecture implements Comparable<Lecture> {
        public int i;
        public long start, end;

        @Override
        public int compareTo(Lecture o) {
            return Long.compare(start, o.start);
        }
    }
}
